pkgs/test_core/lib/src/runner/vm/test_compiler.dart

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:test_api/backend.dart'; // ignore: deprecated_member_use
+import 'package:frontend_server_client/frontend_server_client.dart';
+
+import '../package_version.dart';
+import '../../util/package_config.dart';
+
+/// A request to the [TestCompiler] for recompilation.
+class _CompilationRequest {
+  _CompilationRequest(this.mainUri, this.result, this.metadata);
+
+  Uri mainUri;
+  Metadata metadata;
+  Completer<CompilationResponse> result;
+}
+
+class CompilationResponse {
+  final String? compilerOutput;
+  final int errorCount;
+  final Uri? kernelOutputUri;
+
+  CompilationResponse(
+      {this.compilerOutput, this.errorCount = 0, this.kernelOutputUri});
+}
+
+class TestCompiler {
+  TestCompiler(this._dillCachePath)
+      : _outputDillDirectory =
+            Directory.systemTemp.createTempSync('dart_test.') {
+    _outputDill = File(p.join(_outputDillDirectory.path, 'output.dill'));
+    _compilerController.stream.listen(_onCompilationRequest);
+  }
+
+  final String _dillCachePath;
+  final Directory _outputDillDirectory;
+  final _compilerController = StreamController<_CompilationRequest>();
+  final _compilationQueue = <_CompilationRequest>[];
+
+  late final File _outputDill;
+  FrontendServerClient? _frontendServerClient;
+
+  Future<CompilationResponse> compile(Uri mainDart, Metadata metadata) {
```

Comment:
This could use a doc comment.

Reply:
what would it say?

Reply:
(I don't think for instance `Compiles [mainDart] with [_compiler]` would be a useful comment)

Reply:
Describing the failure behavior would be worthwhile. It could also be useful to describe how this doesn't necessarily start the compilation right away.

```dart
/// Enqueues a request to compile [mainDart] and returns the result.
///
/// This request may need to wait for ongoing compilations.
/// If [dispose] was called, the returned future will immediately complete with `null`.
/// The entrypoint [mainDart] is wrapped in a script to`internalBootstrapVmTest`.
```

I'm not sure that `null` is the right behavior for that - maybe we should throw instead? See my other comment as well about this path getting a test.

Reply:
Changing the behavior to return an actual `CompilerResponse` with an error saying it was shut down.

I also added some comments as suggested.